File: relay/python/relay.py
#!/usr/bin/env python3
"""Minimal TLTV relay — Python 3.9+, single dependency: pip install cryptography"""

import json, os, re, threading, time, urllib.request
from http.server import HTTPServer, BaseHTTPRequestHandler
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
from cryptography.exceptions import InvalidSignature

PORT = int(os.environ.get("PORT", "9000"))
UPSTREAM = os.environ.get("UPSTREAM", "localhost:8000")
BASE = f"http://{UPSTREAM}"

# Base58 (Bitcoin alphabet) — decode needed for signature verification
B58 = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
B58MAP = {c: i for i, c in enumerate(B58)}


def b58decode(s: str) -> bytes:
    n = 0
    for c in s:
        n = n * 58 + B58MAP[c]
    data = n.to_bytes((n.bit_length() + 7) // 8, "big") if n else b""
    return b"\x00" * (len(s) - len(s.lstrip("1"))) + data


def verify_doc(doc: dict) -> bool:
    """Verify Ed25519 signature on a TLTV metadata document — no private key needed."""
    sig, cid = doc.get("signature"), doc.get("id")
    if not sig or not cid:
        return False
    try:
        raw = b58decode(cid)
        if len(raw) < 34 or raw[0] != 0x14 or raw[1] != 0x33:
            return False
        pub = Ed25519PublicKey.from_public_bytes(raw[2:])
        body = {k: v for k, v in doc.items() if k != "signature"}
        canonical = json.dumps(body, sort_keys=True, separators=(",", ":")).encode()
        pub.verify(b58decode(sig), canonical)
        return True
    except (InvalidSignature, Exception):
        return False


def fetch_bytes(path: str) -> bytes:
    return urllib.request.urlopen(f"{BASE}{path}", timeout=5).read()


def fetch_json(path: str):
    return json.loads(fetch_bytes(path))


# Relay state
lock = threading.Lock()
channels = {}  # id -> {meta, manifest, segs: {name: bytes}}


def sync_meta():
    try:
        info = fetch_json("/.well-known/tltv")
        seen = set()
        for ch in [*info.get("channels", []), *info.get("relaying", [])]:
            try:
                meta = fetch_json(f"/tltv/v1/channels/{ch['id']}")
                if not verify_doc(meta):
                    continue
                if meta.get("access") == "token" or meta.get("on_demand"):
                    continue
                cid = ch["id"]
                seen.add(cid)
                with lock:
                    if cid not in channels:
                        channels[cid] = {"meta": meta, "manifest": None, "segs": {}}
                        print(f"  + {meta['name']} ({cid})")
                    elif meta["seq"] > channels[cid]["meta"]["seq"]:
                        if meta.get("status") == "retired":
                            del channels[cid]
                        else:
                            channels[cid]["meta"] = meta
            except Exception:
                pass
        with lock:
            for cid in list(channels):
                if cid not in seen:
                    del channels[cid]
    except Exception as e:
        print(f"  sync: {e}")


def sync_hls():
    with lock:
        items = list(channels.items())
    for cid, ch in items:
        try:
            manifest = fetch_bytes(f"/tltv/v1/channels/{cid}/stream.m3u8")
            names = set()
            for line in manifest.decode().splitlines():
                name = line.strip()
                if name and not name.startswith("#"):
                    names.add(name)
                    if name not in ch["segs"]:
                        try:
                            ch["segs"][name] = fetch_bytes(
                                f"/tltv/v1/channels/{cid}/{name}"
                            )
                        except Exception:
                            pass
            with lock:
                ch["manifest"] = manifest
                for k in list(ch["segs"]):
                    if k not in names:
                        del ch["segs"][k]
        except Exception:
            pass


def poll():
    while True:
        sync_meta()
        for _ in range(30):
            sync_hls()
            time.sleep(2)


# HTTP server
CHAN_RE = re.compile(r"^/tltv/v1/channels/([^/]+)$")
M3U8_RE = re.compile(r"^/tltv/v1/channels/([^/]+)/stream\.m3u8$")
SEG_RE = re.compile(r"^/tltv/v1/channels/([^/]+)/(.+\.ts)$")


class Handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(204)
        self._cors()
        self.end_headers()

    def do_GET(self):
        path = self.path.split("?")[0]

        if path == "/.well-known/tltv":
            with lock:
                relaying = [
                    {"id": c["meta"]["id"], "name": c["meta"]["name"]}
                    for c in channels.values()
                ]
            return self._json(
                {
                    "protocol": "tltv",
                    "versions": [1],
                    "channels": [],
                    "relaying": relaying,
                }
            )

        if path == "/tltv/v1/peers":
            return self._json({"peers": []})

        m = CHAN_RE.match(path)
        if m:
            with lock:
                ch = channels.get(m.group(1))
            if not ch:
                return self._json({"error": "channel_not_found"}, 404)
            return self._json(ch["meta"])

        m = M3U8_RE.match(path)
        if m:
            with lock:
                ch = channels.get(m.group(1))
                manifest = ch["manifest"] if ch else None
            if not ch:
                return self._json({"error": "channel_not_found"}, 404)
            if not manifest:
                return self._json({"error": "stream_unavailable"}, 503)
            self.send_response(200)
            self._cors()
            self.send_header("Content-Type", "application/vnd.apple.mpegurl")
            self.send_header("Cache-Control", "max-age=1, no-cache")
            self.end_headers()
            return self.wfile.write(manifest)

        m = SEG_RE.match(path)
        if m:
            with lock:
                ch = channels.get(m.group(1))
                seg = ch["segs"].get(m.group(2)) if ch else None
            if not ch:
                return self._json({"error": "channel_not_found"}, 404)
            if not seg:
                return self._json({"error": "stream_unavailable"}, 503)
            self.send_response(200)
            self._cors()
            self.send_header("Content-Type", "video/mp2t")
            self.send_header("Cache-Control", "max-age=3600")
            self.end_headers()
            return self.wfile.write(seg)

        self._json({"error": "invalid_request"}, 400)

    def _json(self, data, status=200):
        body = json.dumps(data).encode()
        self.send_response(status)
        self._cors()
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.end_headers()
        self.wfile.write(body)

    def _cors(self):
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET, OPTIONS")

    def log_message(self, *_):
        pass


print(
    f"TLTV Relay  upstream={UPSTREAM}\nListen:  :{PORT}\n\nVerify:  tltv node localhost:{PORT} --local"
)
threading.Thread(target=poll, daemon=True).start()
HTTPServer(("", PORT), Handler).serve_forever()

File: relay/node/relay.mjs
// Minimal TLTV relay — Node.js 18+, zero dependencies
// Usage: UPSTREAM=localhost:8000 node relay.mjs

import { createServer } from 'node:http'
import { verify, createPublicKey } from 'node:crypto'

const PORT = parseInt(process.env.PORT || '9000')
const UPSTREAM = process.env.UPSTREAM || 'localhost:8000'
const UP = `http://${UPSTREAM}`

// Base58 (Bitcoin alphabet) — decode needed for signature verification
const B58 = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz'
const B58MAP = Object.fromEntries([...B58].map((c, i) => [c, BigInt(i)]))

function b58decode(s) {
  let n = 0n
  for (const c of s) n = n * 58n + B58MAP[c]
  const hex = n === 0n ? '' : n.toString(16)
  const padded = hex.length % 2 ? '0' + hex : hex
  const bytes = Buffer.from(padded, 'hex')
  let z = 0
  while (z < s.length && s[z] === '1') z++
  return z ? Buffer.concat([Buffer.alloc(z), bytes]) : bytes
}

// RFC 8785 canonical JSON (handles TLTV types: strings, ints, bools, arrays, objects)
function cjson(v) {
  if (typeof v === 'string') return JSON.stringify(v)
  if (typeof v !== 'object' || v === null) return String(v)
  if (Array.isArray(v)) return '[' + v.map(cjson).join(',') + ']'
  return '{' + Object.keys(v).sort().map(k => JSON.stringify(k) + ':' + cjson(v[k])).join(',') + '}'
}

// Ed25519 signature verification — no private key needed
const SPKI = Buffer.from('302a300506032b6570032100', 'hex')
function verifyDoc(doc) {
  try {
    const id = b58decode(doc.id)
    if (id.length < 34 || id[0] !== 0x14 || id[1] !== 0x33) return false
    const pub = createPublicKey({ key: Buffer.concat([SPKI, id.subarray(2)]), format: 'der', type: 'spki' })
    const body = {}
    for (const k of Object.keys(doc)) if (k !== 'signature') body[k] = doc[k]
    return verify(null, Buffer.from(cjson(body)), pub, b58decode(doc.signature))
  } catch { return false }
}

// Relay state: channelId -> { meta, manifest, segs: Map<name, Buffer> }
const channels = new Map()

async function syncMeta() {
  try {
    const info = await (await fetch(`${UP}/.well-known/tltv`)).json()
    const seen = new Set()
    for (const ch of [...(info.channels || []), ...(info.relaying || [])]) {
      try {
        const meta = await (await fetch(`${UP}/tltv/v1/channels/${ch.id}`)).json()
        if (!verifyDoc(meta)) continue
        if (meta.access === 'token' || meta.on_demand) continue
        seen.add(ch.id)
        const cur = channels.get(ch.id)
        if (!cur) {
          channels.set(ch.id, { meta, manifest: null, segs: new Map() })
          console.log(`  + ${meta.name} (${ch.id})`)
        } else if (meta.seq > cur.meta.seq) {
          if (meta.status === 'retired') { channels.delete(ch.id); continue }
          cur.meta = meta
        }
      } catch {}
    }
    for (const id of channels.keys()) if (!seen.has(id)) channels.delete(id)
  } catch (e) { console.error(`  sync: ${e.message}`) }
}

async function syncHLS() {
  for (const [id, ch] of channels) {
    try {
      ch.manifest = Buffer.from(await (await fetch(`${UP}/tltv/v1/channels/${id}/stream.m3u8`)).arrayBuffer())
      const names = new Set()
      for (const line of ch.manifest.toString().split('\n')) {
        const name = line.trim()
        if (name && !name.startsWith('#')) {
          names.add(name)
          if (!ch.segs.has(name))
            try { ch.segs.set(name, Buffer.from(await (await fetch(`${UP}/tltv/v1/channels/${id}/${name}`)).arrayBuffer())) } catch {}
        }
      }
      for (const k of ch.segs.keys()) if (!names.has(k)) ch.segs.delete(k)
    } catch {}
  }
}

// HTTP server
function json(res, data, status = 200) {
  res.writeHead(status, { 'Content-Type': 'application/json; charset=utf-8' })
  res.end(JSON.stringify(data))
}

createServer((req, res) => {
  res.setHeader('Access-Control-Allow-Origin', '*')
  res.setHeader('Access-Control-Allow-Methods', 'GET, OPTIONS')
  if (req.method === 'OPTIONS') { res.writeHead(204); return res.end() }
  if (req.method !== 'GET') return json(res, { error: 'invalid_request' }, 400)

  const path = new URL(req.url, `http://${req.headers.host}`).pathname

  if (path === '/.well-known/tltv')
    return json(res, { protocol: 'tltv', versions: [1], channels: [],
      relaying: [...channels.values()].map(c => ({ id: c.meta.id, name: c.meta.name })) })
  if (path === '/tltv/v1/peers') return json(res, { peers: [] })

  const m = path.match(/^\/tltv\/v1\/channels\/([^/]+)(\/.*)?$/)
  if (!m) return json(res, { error: 'invalid_request' }, 400)
  const ch = channels.get(m[1])
  if (!ch) return json(res, { error: 'channel_not_found' }, 404)

  if (!m[2]) return json(res, ch.meta)
  if (m[2] === '/stream.m3u8') {
    if (!ch.manifest) return json(res, { error: 'stream_unavailable' }, 503)
    res.writeHead(200, { 'Content-Type': 'application/vnd.apple.mpegurl', 'Cache-Control': 'max-age=1, no-cache' })
    return res.end(ch.manifest)
  }
  const seg = m[2].match(/^\/(.+\.ts)$/)
  if (seg && ch.segs.has(seg[1])) {
    res.writeHead(200, { 'Content-Type': 'video/mp2t', 'Cache-Control': 'max-age=3600' })
    return res.end(ch.segs.get(seg[1]))
  }
  json(res, { error: 'stream_unavailable' }, 503)
}).listen(PORT, async () => {
  console.log(`TLTV Relay  upstream=${UPSTREAM}`)
  console.log(`Listen:  :${PORT}`)
  console.log(`\nVerify:  tltv node localhost:${PORT} --local`)
  await syncMeta()
  setInterval(syncMeta, 60_000)
  setInterval(syncHLS, 2_000)
})
